docs/proposals/systemd-nodespec.md

```diff
@@ -0,0 +1,134 @@
+properties via the ```systemd``` API.  Unit file changes for resource controls are applied automatically on write so no restart is needed.
+The existing ```kubelet``` code that creates a ```/kubelet``` and ```/docker``` cgroup hierarchy off the root should no longer be used on
+```systemd``` systems in favor delegation to the launching processes ```service``` and the ```docker.service``` file.  Direct manipulation
+of the ```docker``` daemon's resource controls should also happen via ```systemd``` to limit the resource growth of docker.
+
+## cgroup per quality of service tier and cgroup parent
+
+```kubelet``` today supports a ```cgroup-root``` flag to state what the root of the cgroup to use for all pods should be.
+
+This flag should be marked deprecated in favor of cgroup per quality of service tiers.
+
+To enable a cgroup per each qos tier, a new flag will be introduced
+
+```
+--cgroup-qos="true": If true, container runtimes are informed to place containers in a cgroup per qos tier.
+```
+
+The ```kubelet``` needs to define a contract for telling the ```DockerManager``` what cgroup to launch a container in based
+upon the configured ```ContainerManager``` implementation.  A generalized interface for a ```CgroupLocator``` that can return
+the fully-qualified cgroup parent path for each quality of service tier will be introduced to fill this need.
+
+## system-container cgroups
+
+The kubelet has the following flag today:
+
+```
+--system-container="": Optional resource-only container in which to place all non-kernel processes that are not already in a container. Empty for no container. Rolling back the flag requires a reboot. (Default: "").
+```
+
+It is proposed that this flag is not supported on ```systemd``` environments.
+
+Instead, the ```kubelet``` will monitor the appropriate slices to monitor resource usage.
+
+There are two options for this:
+
+* in the near term, we could just enable the ```CPUController``` and ```MemoryController``` for the launching node process and the ```docker``` daemon.
+* or we could just turn on both controllers at the ```system.slice``` level and monitor that slice directly to determine host usage.
+
+## monitoring cgroup usage
+
+There are no changes required or proposed in this document.
+
+cAdvisor will continue to be used to monitor resource usage per cgroup.
+
+## downstream providers that wrap ```kubelet```
+
+OpenShift ships a ```openshift-node.service``` that aggregates the ```kubelet``` and ```kube-proxy``` into a single daemon.
+
+When working in that environment, the ```kubelet``` should not make assumptions about the naming of its pid.
+
+For ```systemd``` integration, the ```kubelet``` can control resource controls on its own process by discovering the unit that it was launched from.
+
+```
+pid := os.Getpid()
+read /proc/pid/cgroup
+parse .service unit from 1:name=systemd:/sliceXYZ/node-agent.service
+```
```

Comment:
You can avoid the `os.Getpid()` by reading in `/proc/self/cgroup`.

rkt has [GetOwnCgroupPath](https://github.com/coreos/rkt/blob/master/common/cgroup/cgroup.go#L208) to parse that. It should be possible to reuse the code from that package.

Note that with unified hierarchy, the parser would need to find the line `0::/path` instead of the line `1:name=systemd:/path`.